test/puppetlabs/trapperkeeper/optional_deps_test.clj

```diff
@@ -0,0 +1,187 @@
+        (is (= ["imagine a sonnet"
+                "about designing futures"
+                "rhyming"
+                "is overrated"]
+               (get-sonnet (tka/get-service app :PoetryService))))))))
+
+(deftest call-service-fn-test
+  (testing "when calling into potentially unincluded services"
+    (let [poetry-service (service PoetryService
+                                  {:required []
+                                   :optional [HaikuService SonnetService]}
+                                  (get-haiku [this]
+                                             (tks/call-service-fn this
+                                                                  HaikuService :haiku
+                                                                  ["tea leaves thwart those who"]
+                                                                  ["imagine the saddest haiku"]))
+                                  (get-sonnet [this]
+                                              (tks/call-service-fn this
+                                                                   SonnetService :sonnet
+                                                                   ["designing futures"
+                                                                    ["rhyming" "is overrated"]]
+                                                                   ["imagine the saddest sonnet"])))]
+
+      (testing "and the service has not been included"
+        (let [app (build-app [poetry-service haiku-service] {})]
+          (is (= ["imagine the saddest sonnet"]
+                 (get-sonnet (tka/get-service app :PoetryService))))
+          (is (= ["here is a haiku"
+                  "about the topic you want"
+                  "tea leaves thwart those who"]
+                 (get-haiku (tka/get-service app :PoetryService))))))
+
+      (testing "and the service has been included"
+        (let [app (build-app [poetry-service haiku-service sonnet-service] {})]
+          (is (= ["imagine a sonnet"
+                  "about designing futures"
+                  "rhyming"
+                  "is overrated"]
+                 (get-sonnet (tka/get-service app :PoetryService))))
+          (is (= ["here is a haiku"
+                  "about the topic you want"
+                  "tea leaves thwart those who"]
+                 (get-haiku (tka/get-service app :PoetryService))))))
+
+      (testing "and the named function does not exist"
+        (let [bad-poetry-service (service PoetryService
+                                          {:required []
+                                           :optional [HaikuService SonnetService]}
+                                          (get-haiku [this]
+                                                     (tks/call-service-fn this
+                                                                          HaikuService
+                                                                          :sestina
```

Comment:
This example is too trivial.  Instead of just using `:sestina` you should actually write one.  :trollface: